### PR TITLE
Pass custom request headers when following redirects

### LINKED ISSUE #88: Custom headers not re-used when following redirect
By default the Browser object follows HTTP redirect codes. 

However, it seems that when doing so, any custom headers that were used on the initial request are not re-used when following the redirect. Not sure whether this should be default HTTP behaviour (but I think so?)

I suppose re-using the headers should be done here https://github.com/clue/php-buzz-react/blob/ba4eb355e9ffe4bb2a9071ccc69698d48da548dc/src/Io/Transaction.php#L136

Not sure, but maybe something like: 

```php
$request = $this->messageFactory->request($method, $location, $request->getHeaders());
```

The reason how I found out: we use custom headers to simulate the "Google Bot" by using a UserAgent header like `Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)`

After a redirect, the default UserAgent seems to be used: `React/alpha` defined at https://github.com/reactphp/http-client/blob/d779a3b3c91ee19742839df7395b366da4471d12/src/RequestData.php#L29

### PR DESCRIPTION
**Pull request by seregazhuk**
This PR fixes #88

### CODE REVIEW

**Comment by clue** [comment]
Thank you for this PR, I would love to get this feature in! :+1: 

However, there are a number of HTTP request headers that probably should not be forwarded, for example https://github.com/request/request/commit/210b326fd8625f358e06c59dc11e74468b1de515#diff-ccc0734f65dd7a299409ff07d35be095R1255. Did you have a chance to look into this?

**Comment by seregazhuk** [comment]
@clue done!

I've moved redirect request creation logic into a separate method, so `onResponseRedirect()` handler still stays clear.
For each use case from your suggestion I've added its own test. Maybe it looks too verbose, but you can be sure that this logic is covered. 

Looking forward to your feedback 😎

File: src/Io/Transaction.php [hunk 1]
```diff
@@ -144,6 +142,29 @@ private function onResponseRedirect(ResponseInterface $response, RequestInterfac
...
+     */
+    private function makeRedirectRequest(RequestInterface $request, UriInterface $location)
+    {
+        $originalHost = $request->getUri()->getHost();
+        $request = $request
+            ->withoutHeader('Host')
+            ->withoutHeader('Content-Type')
+            ->withoutHeader('Content-Length');
+
+        if($location->getHost() !== $originalHost) {
```
Comment (clue) [inline src/Io/Transaction.php]: It's my understanding this should probably verify the whole authority, no? (For example when redirecting from https:// to http:// or using another port etc.)

Also, this block could use a small comment (*why*), plus minor CS fix :+1:

Comment (seregazhuk) [reply src/Io/Transaction.php]: We remove authorization only on changing hostname  (not if just changing ports or protocols). I've added a description comment and fixed CS.

**Comment by seregazhuk** [comment]
I've added some comments explaining the logic, and fixes CS issue 😎

File: src/Io/Transaction.php [hunk 2]
```diff
@@ -144,6 +142,30 @@ private function onResponseRedirect(ResponseInterface $response, RequestInterfac
...
+    private function makeRedirectRequest(RequestInterface $request, UriInterface $location)
+    {
+        $originalHost = $request->getUri()->getHost();
+        $request = $request
+            ->withoutHeader('Host')
+            ->withoutHeader('Content-Type')
+            ->withoutHeader('Content-Length');
+
+        // Remove authorization if changing hostnames (but not if just changing ports or protocols).
+        if ($location->getHost() !== $originalHost) {
```
Comment (clue) [inline src/Io/Transaction.php]: See my above comment, do you have a specific use case in mind where it makes sense to keep this if the URI components change, but the host stays the same? It's still my understanding that this should check the URI authority instead, but perhaps I'm missing a relevant use case?

Comment (seregazhuk) [reply src/Io/Transaction.php]: Actually my assumptions about this behavior are based on other implementations:
- https://github.com/request/request/pull/1184
- https://github.com/composer/composer/issues/6716

It looks like `request` and `composer` only check for host difference. Or I'm missing some-thing?

Comment (clue) [reply src/Io/Transaction.php]: Thank you for looking this up, no objections in this case 👍

**Pull request merged**